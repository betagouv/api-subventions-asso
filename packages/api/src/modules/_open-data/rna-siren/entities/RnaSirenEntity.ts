import { Rna, Siren } from "dto";

export default class RnaSiren {
    constructor(public rna: Rna, public siren: Siren, public names?: string[]) {}
}
