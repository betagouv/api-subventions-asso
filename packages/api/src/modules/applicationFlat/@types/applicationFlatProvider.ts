import { FindCursor } from "mongodb";
import Provider from "../../providers/@types/IProvider";
import ApplicationFlatEntity from "../../../entities/ApplicationFlatEntity";

export default interface ApplicationFlatProvider extends Provider {
    isApplicationFlatProvider: true;

    findAllByExercise(exercise?: number): FindCursor;

    getApplicationFlatStream(START_YEAR: number, END_YEAR: number): ReadableStream<ApplicationFlatEntity>;
}
