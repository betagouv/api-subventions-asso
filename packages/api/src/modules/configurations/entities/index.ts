export * from "./DauphinTokenAvailableTimeDataEntity";
export * from "./DauphinTokenDataEntity";
export * as ConfigurationEntity from "./ConfigurationEntity";
