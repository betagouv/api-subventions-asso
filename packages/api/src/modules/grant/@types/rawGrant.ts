export type RawGrant = {
    provider: string;
    type: "application" | "fullGrant" | "payment";
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- neither never nor unknown works
    data: Record<string, any>;
    joinKey?: string;
};

export type JoinedRawGrant = {
    fullGrant?: RawGrant;
    payments?: RawGrant[];
    application?: RawGrant;
};
