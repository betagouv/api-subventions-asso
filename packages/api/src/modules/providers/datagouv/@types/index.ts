export * from "./ISaveCallbacks";
export * from "./UniteLegalRaw";
