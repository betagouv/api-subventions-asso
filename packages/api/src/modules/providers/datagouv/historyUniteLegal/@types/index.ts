export * from "./ISaveCallbacks";
export * from "./UniteLegalHistoryRow"