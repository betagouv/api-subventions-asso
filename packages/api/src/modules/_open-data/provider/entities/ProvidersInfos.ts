import { Provider } from "../@types/Provider";

export default class ProvidersInfos {
    constructor(public api: Provider[], public raw: Provider[]) {}
}
