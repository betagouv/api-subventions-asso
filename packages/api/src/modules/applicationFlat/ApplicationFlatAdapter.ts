import { DemandeSubvention } from "dto";
import { WithId } from "mongodb";
import { ApplicationFlatEntity } from "../../entities/ApplicationFlatEntity";
import { RawApplication } from "../grant/@types/rawGrant";
import ProviderValueAdapter from "../../shared/adapters/ProviderValueAdapter";
import applicationFlatService from "./applicationFlat.service";

// entities and draft are almost equal but we want ids to be built in constructor
// and we want to be able to build with a properly types object
type ApplicationFlatEntityDraft = Omit<ApplicationFlatEntity, "idUnique" | "idSubvention">;

export default class ApplicationFlatAdapter {
    public static rawToApplication(rawApplication: RawApplication<ApplicationFlatEntity>) {
        return this.toDemandeSubvention(rawApplication.data);
    }

    public static buildEntity(draft: ApplicationFlatEntityDraft): ApplicationFlatEntity {
        return {
            ...draft,
            idSubvention: `${draft.provider}--${draft.idSubventionProvider}`,
            idUnique: `${draft.provider}--${draft.idSubventionProvider}}--${draft.exerciceBudgetaire}`,
        };
    }

    public static dboToEntity(dbo: WithId<ApplicationFlatEntity>) {
        const { _id, ...entity } = dbo;
        return entity;
    }

    public static toDemandeSubvention(entity: ApplicationFlatEntity): DemandeSubvention | null {
        const siret = applicationFlatService.getSiret(entity);
        if (!siret) return null;

        const toPv = <T>(value: T) =>
            ProviderValueAdapter.toProviderValue<T>(value, entity.provider, entity.dateConvention); // TODO bad date

        const toPvOrUndefined = value => (value ? toPv(value) : undefined);

        /* Pour l'instant on garde ej pour tous les providers sauf Fonjep qui prend idVersement 
        Il faudra convertir tous les versementKey en idVersement quand tout est connect√©  */
        return {
            annee_demande: toPvOrUndefined(entity.anneeDemande),
            date_commision: toPvOrUndefined(entity.dateDecision), // TODO surely not good
            pluriannualite: toPvOrUndefined(entity.pluriannualite),
            service_instructeur: toPv(entity.nomServiceInstructeur || ""),
            siret: toPv(siret.value),
            sous_dispositif: toPvOrUndefined(entity.sousDispositif),
            status: toPv(entity.statutLabel || ""),
            statut_label: toPv(entity.statutLabel),
            transmis_le: toPvOrUndefined(entity.dateDepotDemande),
            versementKey: toPvOrUndefined(entity.idVersement),
            ej: toPvOrUndefined(entity.ej),
            creer_le: toPvOrUndefined(entity.dateDepotDemande),
            dispositif: toPvOrUndefined(entity.dispositif),
            montants: {
                total: toPvOrUndefined(entity.montantTotal),
                demande: toPvOrUndefined(entity.montantDemande),
                accorde: toPvOrUndefined(entity.montantAccorde),
            },
            financeur_principal: toPvOrUndefined(entity.nomAttribuant), // TODO
            actions_proposee:
                entity.objet === "Fonctionnement global" || entity.objet == undefined
                    ? undefined
                    : [{ intitule: toPv(entity.objet) }],
            co_financement: {
                cofinanceur: toPv(entity.nomsAttribuantsCofinanceurs?.join(", ") || ""),
                cofinanceur_email: toPv(""),
                montants: toPv(0), // TODO fake data but we won't know
            },
        };
    }
}
