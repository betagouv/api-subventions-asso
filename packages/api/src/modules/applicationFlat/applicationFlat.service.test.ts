import Siren from "../../identifierObjects/Siren";

import applicationFlatPort from "../../dataProviders/db/applicationFlat/applicationFlat.port";
import applicationFlatService from "./applicationFlat.service";
import { ApplicationFlatEntity } from "../../entities/ApplicationFlatEntity";
import ApplicationFlatAdapter from "./ApplicationFlatAdapter";
import { DemandeSubvention } from "dto";
import { RawApplication } from "../grant/@types/rawGrant";
import { ReadableStream } from "node:stream/web";

import Siret from "../../identifierObjects/Siret";
import AssociationIdentifier from "../../identifierObjects/AssociationIdentifier";
import EstablishmentIdentifier from "../../identifierObjects/EstablishmentIdentifier";
import { FindCursor } from "mongodb";
import { insertStreamByBatch } from "../../shared/helpers/MongoHelper";

jest.mock("../../dataProviders/db/applicationFlat/applicationFlat.port");
jest.mock("./ApplicationFlatAdapter");
jest.mock("../../identifierObjects/Siret");
jest.mock("../../shared/helpers/MongoHelper");

describe("ApplicationFlatService", () => {
    const FLAT_ENTITY = {} as ApplicationFlatEntity;
    describe("application part", () => {
        const [E1, E2] = ["entity1", "entity2"] as unknown as ApplicationFlatEntity[];
        describe.each`
            identifierType | rawIdentifier                  | findMethod                         | identifierConstructor
            ${"siret"}     | ${new Siret("12345678901234")} | ${applicationFlatPort.findBySiret} | ${EstablishmentIdentifier.fromSiret}
            ${"siren"}     | ${new Siren("123456789")}      | ${applicationFlatPort.findBySiren} | ${AssociationIdentifier.fromSiren}
        `("getEntitiesByIdentifier", ({ rawIdentifier, identifierConstructor, findMethod }) => {
            beforeAll(() => {
                findMethod.mockResolvedValue([E1, E2]);
            });
            afterAll(() => {
                findMethod.mockRestore();
            });

            it("if identifier is $identifierType, call proper port method", async () => {
                await applicationFlatService.getEntitiesByIdentifier(identifierConstructor(rawIdentifier));
                expect(findMethod).toHaveBeenCalledWith(rawIdentifier);
            });

            it("returns entities", async () => {
                const expected = [E1, E2];
                const actual = await applicationFlatService.getEntitiesByIdentifier(
                    identifierConstructor(rawIdentifier),
                );
                expect(actual).toEqual(expected);
            });
        });

        describe("rawToApplication", () => {
            const RAW_GRANT = {
                type: "application",
                provider: "some",
                data: FLAT_ENTITY,
                joinKey: "ej",
            } as RawApplication<ApplicationFlatEntity>;

            it("calls adapter", () => {
                applicationFlatService.rawToApplication(RAW_GRANT);
                expect(ApplicationFlatAdapter.rawToApplication).toHaveBeenCalledWith(RAW_GRANT);
            });

            it("returns adapter's result", () => {
                const expected = "adapted" as unknown as DemandeSubvention;
                jest.mocked(ApplicationFlatAdapter.rawToApplication).mockReturnValueOnce(expected);
                const actual = applicationFlatService.rawToApplication(RAW_GRANT);
                expect(actual).toBe(expected);
            });
        });

        describe("getDemandeSubvention", () => {
            let getEntitiesSpy;
            const [E1, E2] = ["entity1", "entity2"] as unknown as ApplicationFlatEntity[];
            const IDENTIFIER = AssociationIdentifier.fromSiren(new Siren("987654321"));

            beforeAll(() => {
                getEntitiesSpy = jest
                    .spyOn(applicationFlatService, "getEntitiesByIdentifier")
                    .mockResolvedValue([E1, E2]);
            });
            afterAll(() => {
                getEntitiesSpy.mockRestore();
            });

            it("gets entities", async () => {
                await applicationFlatService.getDemandeSubvention(IDENTIFIER);
                expect(getEntitiesSpy).toHaveBeenCalledWith(IDENTIFIER);
            });

            it("adapts all applications", async () => {
                await applicationFlatService.getDemandeSubvention(IDENTIFIER);
                expect(ApplicationFlatAdapter.toDemandeSubvention).toHaveBeenCalledWith(E1);
                expect(ApplicationFlatAdapter.toDemandeSubvention).toHaveBeenCalledWith(E2);
            });

            it("returns non-null adapted applications", async () => {
                const A2 = "adapted 2" as unknown as DemandeSubvention;
                jest.mocked(ApplicationFlatAdapter.toDemandeSubvention).mockReturnValueOnce(null);
                jest.mocked(ApplicationFlatAdapter.toDemandeSubvention).mockReturnValue(A2);
                const expected = [A2];
                const actual = await applicationFlatService.getDemandeSubvention(IDENTIFIER);
                expect(actual).toEqual(expected);
            });
        });
    });

    describe("grant part", () => {
        describe("getRawGrants", () => {
            let getEntitiesSpy;
            const [E1, E2] = [
                { provider: "fonjep", paymentId: "poste1" },
                { provider: "autre", ej: "ej2" },
            ] as unknown as ApplicationFlatEntity[];
            const IDENTIFIER = AssociationIdentifier.fromSiren(new Siren("987654321"));

            beforeAll(() => {
                getEntitiesSpy = jest
                    .spyOn(applicationFlatService, "getEntitiesByIdentifier")
                    .mockResolvedValue([E1, E2]);
            });
            afterAll(() => {
                getEntitiesSpy.mockRestore();
            });

            it("gets entities", async () => {
                await applicationFlatService.getRawGrants(IDENTIFIER);
                expect(getEntitiesSpy).toHaveBeenCalledWith(IDENTIFIER);
            });

            it("converts found methods", async () => {
                const actual = await applicationFlatService.getRawGrants(IDENTIFIER);
                expect(actual).toMatchSnapshot();
            });
        });
    });

    describe("applicationFlat part", () => {
        describe("saveFromStream", () => {
            const STREAM = {} as unknown as ReadableStream;
            it("calls mongo helper", async () => {
                await applicationFlatService.saveFromStream(STREAM);
                expect(insertStreamByBatch).toHaveBeenCalledWith(STREAM, applicationFlatPort.upsertMany, 10000);
            });
        });
    });

    describe("isCollectionInitialized", () => {
        it("calls check in port", () => {
            applicationFlatService.isCollectionInitialized();
            expect(applicationFlatPort.hasBeenInitialized).toHaveBeenCalled();
        });
    });

    describe("getSiret", () => {
        beforeAll(() => {
            jest.mocked(Siret.isSiret).mockReturnValueOnce(false);
        });
        afterAll(() => {
            jest.mocked(Siret.isSiret).mockRestore();
        });

        const ENTITY = { idBeneficiaire: "123456789", typeIdBeneficiaire: "siret" } as unknown as ApplicationFlatEntity;
        it("returns undefined if typeIdBeneficiaire is not siret", () => {
            const actual = applicationFlatService.getSiret(ENTITY);
            expect(actual).toBeUndefined();
        });

        it("returns undefined if not siret", () => {
            jest.mocked(Siret.isSiret).mockReturnValueOnce(false);
            const actual = applicationFlatService.getSiret(ENTITY);
            expect(actual).toBeUndefined();
        });

        it("returns valueObject from entity", () => {
            const actual = applicationFlatService.getSiret(ENTITY);
            expect(actual).toMatchInlineSnapshot(`undefined`);
        });
    });

    describe("containsDataFromProvider", () => {
        const CURSOR = { hasNext: jest.fn() } as unknown as FindCursor;
        const PROVIDER = "PROV";

        beforeAll(() => {
            jest.mocked(applicationFlatPort.cursorFind).mockReturnValue(CURSOR);
        });

        afterAll(() => {
            jest.mocked(applicationFlatPort.cursorFind).mockRestore();
        });

        it("gets cursor", async () => {
            await applicationFlatService.containsDataFromProvider(PROVIDER);
            expect(applicationFlatPort.cursorFind({ provider: PROVIDER }));
        });

        it("returns response from cursor's hasNext", async () => {
            const expected = "toto" as unknown as boolean;
            jest.mocked(CURSOR.hasNext).mockResolvedValue(expected);
            const actual = await applicationFlatService.containsDataFromProvider(PROVIDER);
            expect(CURSOR.hasNext).toHaveBeenCalled();
            expect(actual).toBe(expected);
        });
    });
});
