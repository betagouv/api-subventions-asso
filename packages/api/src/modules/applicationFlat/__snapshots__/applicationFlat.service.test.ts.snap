// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationFlatService grant part getRawGrants converts found methods 1`] = `
[
  {
    "data": {
      "idVersement": "poste1",
      "provider": "fonjep",
    },
    "joinKey": "poste1",
    "provider": "fonjep",
    "type": "application",
  },
  {
    "data": {
      "ej": "ej2",
      "provider": "autre",
    },
    "joinKey": "ej2",
    "provider": "autre",
    "type": "application",
  },
]
`;
